lib/src/command/outdated.dart

```diff
@@ -300,6 +301,72 @@ class OutdatedCommand extends PubCommand {
       return [];
     }
   }
+
+  Future<List<List<_FormattedString>>> nullSafetyMarker(
+      List<_PackageDetails> packages) async {
+    final nullSafetyMap = await log.spinner('Computing null-safety', () async {
+      /// Find all unique ids.
+      final ids = {
+        for (final packageDetails in packages) ...[
+          packageDetails.current?._id,
+          packageDetails.upgradable?._id,
+          packageDetails.resolvable?._id,
+          packageDetails.latest?._id,
+        ]
+      }.where((id) => id != null);
+      final nullSafetyAnalyzer = NullSafetyAnalysis(cache);
+      return Map.fromEntries(
+        await Future.wait(
+          ids.map(
+            (id) async => MapEntry(
+              id,
+              await nullSafetyAnalyzer.nullSafetyCompliance(id),
+            ),
+          ),
+        ),
+      );
```

Comment:
Pretty awesome, but... `NullSafetyAnalysis.nullSafetyCompliance` may install packages into `PUB_CACHE`.

I'm just now realizing that running `pub get` concurrently isn't safe, well, maybe... I'm not entirely sure. We probably could make it safe within the same process without much trouble.

If you look at this snippet from `_installInCache` used by `pub global activate`

https://github.com/dart-lang/pub/blob/6deb457048deb435009b36a4cd2d86003d107cf4/lib/src/global_packages.dart#L197-L205

For each resolution we make, we have a `result` object, where we need to iterate over `result.packages`. But we don't want to call `await source.downloadToSystemCache(id);` for the same package-version more than once.

Also if we look at `BoundHostedSource.downloadToSystemCache`:

https://github.com/dart-lang/pub/blob/6deb457048deb435009b36a4cd2d86003d107cf4/lib/src/source/hosted.dart#L258-L268

It's clear that we only download it if it does not already exist in `PUB_CACHE` this is checked by `isInSystemCache(id)`, which simply checks if the folder exists. We also create the folder before we download.
So if there is a race condition, we might download it twice, this is unlikely. However, it's fairly like that the first time we call `downloadToSystemCache` it creates the directory and starts downloading, and the next time we call it (while it's still being downloaded) `downloadToSystemCache` will return fast because the directory exists, even though the package is in fact not done being downloaded. Hence, when running analysis we may be missing some or all of the source files because they are still being downloaded.

-----------

I think we could remedy this by refactoring `NullSafetyAnalysis` such that `NullSafetyAnalysis.enqueueAnalysis(id)` is an async function that returns a Future for the result, but doesn't start processing until `NullSafetyAnalysis.populateCache()` has been called.

Inside `NullSafetyAnalysis.enqueueAnalysis(id)` we run the dependency resolution, then we add a tuple consisting of a `Completer` and a list of package ids for which we want `downloadToSystemCache` to be called.. And we await the completer being done before starting the analysis.

Then in `populateCache()` we run through the list of all tuples of completers and packages ids to be downloaded to cache, and we remove duplicate package ids and download them to cache before resolving the completer.

This is a bit crazy and there is some non-trivial synchronization going on.

Reply:
ah, looks like `_packageInternallyGoodCache` saves us here. It might be nice to document why this is...

Reply:
> NullSafetyAnalysis.nullSafetyCompliance may install packages into PUB_CACHE.

I thought we agreed this was acceptable offline. We might be able to install unused versions somewhere else and delete them afterwards.

> downloadToSystemCache will return fast because the directory exists,

That's a good point.

I think easier will be a map keeping futures of all the packages we install. Then we can use putIfAbsent to populate it, and each package will go in only once:

```dart
class NullSafetyAnalyzer {
final Map<PackageId, Future<void>> _packagesToBeInstalled = {};
...

Future nullSafetyCompliance() {
   ...
  // instead of `await boundSource.downloadToSystemCache(dependencyId);`
  await _packagesToBeInstalled.putIfAbsent(dependencyId, () =>
   boundSource.downloadToSystemCache(dependencyId);
  ---
}
```

Reply:
Installing into cache is fine, my concern was concurrency.